vaults/azurekeyvault.go

```diff
@@ -0,0 +1,139 @@
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+)
+
+// AKVSecretConfig provides the options to get secret from Azure keyvault using MSI.
+type AKVSecretConfig struct {
+	VaultURL       string
+	SecretName     string
+	SecretVersion  string
+	MSIClientID    string
+	AADResourceURL string
+}
+
+// GetValue gets the secret vaule as defined by the config from Azure key vault using MSI.
+func (secretConfig *AKVSecretConfig) GetValue() (string, error) {
+	if secretConfig == nil {
+		return "", errors.New("secret config is required")
+	}
+
+	if secretConfig.VaultURL == "" ||
+		secretConfig.SecretName == "" {
+		return "", errors.New("missing required properties vaultURL and SecretName")
+	}
+
+	keyClient, err := newKeyVaultClient(secretConfig.VaultURL, secretConfig.MSIClientID, secretConfig.AADResourceURL)
+	if err != nil {
+		return "", fmt.Errorf("failed to create azure key vault client. Error: %v", err)
+	}
+
+	secretValue, err := keyClient.getSecret(secretConfig.SecretName, secretConfig.SecretVersion)
+	if err != nil {
+		return "", fmt.Errorf("failed to fetch secret value from azure key vault client. Error: %v", err)
+	}
+
+	return secretValue, nil
+}
+
+// NewAKVSecretConfig creates the Azure Key Vault config.
+func NewAKVSecretConfig(vaultURL, msiClientID, vaultAADResourceURL string) (*AKVSecretConfig, error) {
+	if vaultURL == "" {
+		return nil, fmt.Errorf("missing azure keyvault URL")
+	}
+
+	normalizedVaultURL := strings.TrimSuffix(strings.ToLower(vaultURL), "/")
+
+	url, err := url.Parse(normalizedVaultURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse the azure keyvault secret URL. Error: %v", err)
+	}
+
+	if url.Scheme != "https" {
+		return nil, fmt.Errorf("invalid azure keyvault secret URL scheme. Expected Https")
+	}
+
+	urlSegments := strings.Split(url.Path, "/")
+
+	if len(urlSegments) != 3 && len(urlSegments) != 4 {
+		return nil, fmt.Errorf("invalid azure keyvault secret URL. Bad number of URL segments: %d", len(urlSegments))
+	}
+
+	if urlSegments[1] != "secrets" {
+		return nil, fmt.Errorf("invalid azure keyvault secret URL. Expected 'secrets' collection, but found: %s", urlSegments[1])
+	}
+
+	secretVersion := ""
+
+	if len(urlSegments) == 4 {
+		secretVersion = urlSegments[3]
+	}
+
+	akvConfig := &AKVSecretConfig{
+		VaultURL:       fmt.Sprintf("%s://%s", url.Scheme, url.Host),
+		SecretName:     urlSegments[2],
+		SecretVersion:  secretVersion,
+		MSIClientID:    msiClientID,
+		AADResourceURL: vaultAADResourceURL,
+	}
+
+	return akvConfig, nil
+}
+
+// KeyVault holds the information for a keyvault instance
+type keyVault struct {
+	client   *keyvault.BaseClient
+	vaultURL string
+}
+
+// NewKeyVaultClient creates a new keyvault client
+func newKeyVaultClient(vaultURL, clientID, vaultAADResourceURL string) (*keyVault, error) {
+
+	if vaultAADResourceURL == "" {
+		vaultAADResourceURL = azure.PublicCloud.KeyVaultEndpoint
+	}
+
+	msiKeyConfig := &auth.MSIConfig{
```

Comment:
Does the auth library allow overriding the default msi endpoint? /cc: @shahzzzam

Reply:
@northtyphoon  Unfortunately the go SDK for App auth hard codes the MSI endpoint https://github.com/Azure/go-autorest/blob/71d02b67e17a3d80abf8db171f1ca80c39eb2e90/autorest/adal/token.go#L63 

Are we planning to start token service at a different endpoint?